File: src/sla_check.py
from datetime import datetime, timedelta
import os
import heapq
from collections import defaultdict
from slack_sdk import WebClient
from database import get_prs_from_store, remove_pr_by_id, get_pr_by_id
from database_channel_settings import get_channel_sla_time, get_channel_enabled_hours
from dotenv import load_dotenv
from helpers import get_status, get_username
import logging

load_dotenv()

client = WebClient(token=os.environ['SLACK_BOT_TOKEN'])

convert_seconds = 3600
logger = logging.getLogger(__name__)

def is_within_working_hours(timestamp):
    """Check if the given timestamp is within 9 AM to 5 PM working hours."""
    return 9 <= timestamp.hour < 17 and timestamp.weekday() < 5  # Monday to Friday

def calculate_working_hours(start_time, end_time):
    """Calculate working hours between two times (9 AM to 5 PM, Monday to Friday)."""
    total_seconds = 0
    current = start_time

    while current < end_time:
        if is_within_working_hours(current):
            next_boundary = min(end_time, current.replace(hour=17, minute=0, second=0, microsecond=0))
            total_seconds += (next_boundary - current).total_seconds()

        current += timedelta(days=1)
        current = current.replace(hour=9, minute=0, second=0, microsecond=0)

    return total_seconds

def format_time_overdue(time_overdue_seconds):
    """Convert time overdue from seconds to hours and minutes."""
    minutes, seconds = divmod(time_overdue_seconds, 60)
    hours, minutes = divmod(minutes, 60)
    return f"{int(hours)} hours, {int(minutes)} minutes"

def format_time_until_overdue(time_elapsed_seconds, channel_sla_time):
    """Convert time until overdue from seconds to minutes."""
    channel_sla_seconds = int(channel_sla_time) * 3600
    remaining_seconds = channel_sla_seconds - time_elapsed_seconds
    remaining_minutes = remaining_seconds // 60
    return f"{int(remaining_minutes)} minutes until overdue"

def is_pr_due_for_removal(pr, now):
    """Check if PR is older than 5 working days (excluding weekends)."""
    pr_timestamp = datetime.fromisoformat(pr["timestamp"])
    working_days = 0
    current = pr_timestamp

    while current < now:
        if current.weekday() < 5:  # Monday to Friday are working days
            working_days += 1
        current += timedelta(days=1)

    return working_days >= 5

def populate_sla(pr_store):
    now = datetime(2024, 9, 6, 16, 40)
    prs_to_remove = []
    overdue_pr_heap = defaultdict(list)
    near_sla_prs = defaultdict(list)

    for pr in pr_store:
        pr_id = pr["id"]
        channel_id = pr["channel_id"]
        channel_sla_time = get_channel_sla_time(channel_id)

        if is_pr_due_for_removal(pr, now):
            logger.info(f"PR {pr_id} is older than 5 working days. Removing it.")
            client.chat_update(
                channel=pr["channel_id"],
                ts=pr["message_ts"],
                text=f"PR *<{pr['link']}|{pr['name']}>* has been automatically removed after 5 working days.",
                blocks=[]
            )
            prs_to_remove.append(pr_id)
            continue

        pr_timestamp = datetime.fromisoformat(pr["timestamp"])
        time_elapsed = calculate_working_hours(pr_timestamp, now)
        reviews_needed = pr["reviews_received"] < pr["reviews_needed"]

        if time_elapsed > channel_sla_time * convert_seconds and reviews_needed:
            time_overdue_seconds = time_elapsed - channel_sla_time * convert_seconds
            heapq.heappush(overdue_pr_heap[pr["channel_id"]], (-time_overdue_seconds, pr_id))
        elif (channel_sla_time - 1) * convert_seconds <= time_elapsed <= channel_sla_time * convert_seconds and reviews_needed:
            near_sla_prs[pr["channel_id"]].append((pr, time_elapsed))

    return prs_to_remove, overdue_pr_heap, near_sla_prs

def populate_message_text(channel_id, overdue_pr_heap, near_sla_prs):
    channel_sla_time = get_channel_sla_time(channel_id)
    message_text = "*:mega: PR Review Reminder*\n"
    message_text += f"SLA time for this channel: {channel_sla_time} hours\n"
    message_text += "Here's a summary of PRs that need your attention:\n\n"
    
    if channel_id in overdue_pr_heap:
        message_text += ":warning: *The following PRs are overdue for review*\n\n"
        while overdue_pr_heap[channel_id]:
            time_overdue, pr_id = heapq.heappop(overdue_pr_heap[channel_id])
            pr = get_pr_by_id(pr_id)
            formatted_time_overdue = format_time_overdue(-time_overdue)
            submitter_name = get_username(client, pr['submitter_id'])
            pr_status = get_status(pr)
            message_text += (
                f"• *<{pr['permalink']}|{pr['name']}>* by {submitter_name}\n"
                f"   - Overdue by {formatted_time_overdue}\n"
                f"   - _Status_: {pr_status}\n"
            )
    
    if channel_id in near_sla_prs:
        if overdue_pr_heap:
            message_text += "\n"
        message_text += ":hourglass_flowing_sand: *The following PRs are within 1 hour of SLA*\n\n"
        for pr, time_elapsed in near_sla_prs[channel_id]:
            submitter_name = get_username(client, pr['submitter_id'])
            pr_status = get_status(pr)
            formatted_time_until_overdue = format_time_until_overdue(time_elapsed, channel_sla_time)
            message_text += (
                f"• *<{pr['permalink']}|{pr['name']}>* by {submitter_name}\n"
                f"   - {formatted_time_until_overdue}\n"
                f"   - _Status_: {pr_status}\n"
            )

    return message_text

def check_sla(client):
    """Check all PRs in the store for SLA violations and those within 1 hour of the SLA."""
    now = datetime(2024, 9, 6, 16, 40)
    logger.info(f"Running SLA check at {now}")
    current_hour = now.hour
    pr_store = get_prs_from_store()
    prs_to_remove, overdue_pr_heap, near_sla_prs = populate_sla(pr_store)

    for pr_id in prs_to_remove:
        remove_pr_by_id(pr_id)

    channels_to_notify = set(overdue_pr_heap.keys()).union(near_sla_prs.keys())

    for channel_id in channels_to_notify:
        enabled_hours = get_channel_enabled_hours(channel_id)
        if current_hour not in enabled_hours:
            logger.info(f"Skipping channel {channel_id}, current hour {current_hour} is not within enabled hours.")
            continue

        message_text = populate_message_text(channel_id, overdue_pr_heap, near_sla_prs)
        if message_text:
            client.chat_postMessage(channel=channel_id, text=message_text, unfurl_links=False)
            logger.info(f"Sent SLA notification for channel {channel_id}.")

if __name__ == "__main__":
    check_sla(client)

File: src/SlaCheck.java
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.stream.Collectors;
import com.slack.api.Slack;
import com.slack.api.methods.SlackApiException;
import com.slack.api.methods.response.chat.ChatPostMessageResponse;
import com.slack.api.methods.response.chat.ChatUpdateResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SlaCheck {

    private static final Logger logger = LoggerFactory.getLogger(SlaCheck.class);
    private static final long CONVERT_SECONDS = 3600;

    public static boolean isWithinWorkingHours(LocalDateTime timestamp) {
        return timestamp.getHour() >= 9 && timestamp.getHour() < 17 && timestamp.getDayOfWeek().getValue() <= 5;
    }

    public static long calculateWorkingHours(LocalDateTime startTime, LocalDateTime endTime) {
        long totalSeconds = 0;
        LocalDateTime current = startTime;

        while (current.isBefore(endTime)) {
            if (isWithinWorkingHours(current)) {
                LocalDateTime nextBoundary = current.withHour(17).withMinute(0).withSecond(0).withNano(0);
                nextBoundary = nextBoundary.isAfter(endTime) ? endTime : nextBoundary;
                totalSeconds += Duration.between(current, nextBoundary).getSeconds();
            }
            current = current.plusDays(1).withHour(9).withMinute(0).withSecond(0).withNano(0);
        }
        return totalSeconds;
    }

    public static String formatTimeOverdue(long timeOverdueSeconds) {
        long minutes = (timeOverdueSeconds / 60) % 60;
        long hours = timeOverdueSeconds / 3600;
        return String.format("%d hours, %d minutes", hours, minutes);
    }

    public static String formatTimeUntilOverdue(long timeElapsedSeconds, int channelSlaTime) {
        long remainingSeconds = channelSlaTime * CONVERT_SECONDS - timeElapsedSeconds;
        long remainingMinutes = remainingSeconds / 60;
        return String.format("%d minutes until overdue", remainingMinutes);
    }

    public static boolean isPrDueForRemoval(Map<String, Object> pr, LocalDateTime now) {
        LocalDateTime prTimestamp = LocalDateTime.parse((String) pr.get("timestamp"));
        long workingDays = 0;
        LocalDateTime current = prTimestamp;

        while (current.isBefore(now)) {
            if (current.getDayOfWeek().getValue() <= 5) {
                workingDays++;
            }
            current = current.plusDays(1);
        }
        return workingDays >= 5;
    }

    public static Map<String, Object> populateSla(List<Map<String, Object>> prStore) {
        LocalDateTime now = LocalDateTime.now();
        List<String> prsToRemove = new ArrayList<>();
        Map<String, PriorityQueue<Map.Entry<Long, String>>> overduePrHeap = new HashMap<>();
        Map<String, List<Map.Entry<Map<String, Object>, Long>>> nearSlaPrs = new HashMap<>();

        for (Map<String, Object> pr : prStore) {
            String prId = (String) pr.get("id");
            String channelId = (String) pr.get("channel_id");
            int channelSlaTime = DatabaseChannelSettings.getChannelSlaTime(channelId);

            if (isPrDueForRemoval(pr, now)) {
                logger.info(String.format("PR %s is older than 5 working days. Removing it.", prId));
                Slack.getInstance().methods().chatUpdate(r -> r.channel(channelId).text("PR has been automatically removed after 5 working days."));
                prsToRemove.add(prId);
                continue;
            }

            LocalDateTime prTimestamp = LocalDateTime.parse((String) pr.get("timestamp"));
            long timeElapsed = calculateWorkingHours(prTimestamp, now);
            boolean reviewsNeeded = (int) pr.get("reviews_received") < (int) pr.get("reviews_needed");

            if (timeElapsed > channelSlaTime * CONVERT_SECONDS && reviewsNeeded) {
                long timeOverdueSeconds = timeElapsed - channelSlaTime * CONVERT_SECONDS;
                overduePrHeap.computeIfAbsent(channelId, k -> new PriorityQueue<>(Map.Entry.comparingByKey())).add(Map.entry(-timeOverdueSeconds, prId));
            } else if (timeElapsed >= (channelSlaTime - 1) * CONVERT_SECONDS && timeElapsed <= channelSlaTime * CONVERT_SECONDS && reviewsNeeded) {
                nearSlaPrs.computeIfAbsent(channelId, k -> new ArrayList<>()).add(Map.entry(pr, timeElapsed));
            }
        }

        return Map.of("prsToRemove", prsToRemove, "overduePrHeap", overduePrHeap, "nearSlaPrs", nearSlaPrs);
    }

    public static String populateMessageText(String channelId, Map<String, PriorityQueue<Map.Entry<Long, String>>> overduePrHeap, Map<String, List<Map.Entry<Map<String, Object>, Long>>> nearSlaPrs) {
        int channelSlaTime = DatabaseChannelSettings.getChannelSlaTime(channelId);
        StringBuilder messageText = new StringBuilder("*:mega: PR Review Reminder*\n");
        messageText.append(String.format("SLA time for this channel: %d hours\n", channelSlaTime));
        messageText.append("Here's a summary of PRs that need your attention:\n\n");

        PriorityQueue<Map.Entry<Long, String>> overduePrs = overduePrHeap.get(channelId);
        if (overduePrs != null && !overduePrs.isEmpty()) {
            messageText.append(":warning: *The following PRs are overdue for review*\n\n");
            while (!overduePrs.isEmpty()) {
                Map.Entry<Long, String> overduePr = overduePrs.poll();
                String prId = overduePr.getValue();
                Map<String, Object> pr = Database.getPrById(prId);
                String formattedTimeOverdue = formatTimeOverdue(-overduePr.getKey());
                String submitterName = Utils.getUsername(Slack.getInstance(), (String) pr.get("submitter_id"));
                String prStatus = Utils.getStatus(pr);
                messageText.append(String.format("• *<%s|%s>* by %s\n   - Overdue by %s\n   - _Status_: %s\n", pr.get("permalink"), pr.get("name"), submitterName, formattedTimeOverdue, prStatus));
            }
        }

        List<Map.Entry<Map<String, Object>, Long>> nearSlaList = nearSlaPrs.get(channelId);
        if (nearSlaList != null && !nearSlaList.isEmpty()) {
            messageText.append("\n:hourglass_flowing_sand: *The following PRs are within 1 hour of SLA*\n\n");
            for (Map.Entry<Map<String, Object>, Long> nearSla : nearSlaList) {
                Map<String, Object> pr = nearSla.getKey();
                String submitterName = Utils.getUsername(Slack.getInstance(), (String) pr.get("submitter_id"));
                String prStatus = Utils.getStatus(pr);
                String formattedTimeUntilOverdue = formatTimeUntilOverdue(nearSla.getValue(), channelSlaTime);
                messageText.append(String.format("• *<%s|%s>* by %s\n   - %s\n   - _Status_: %s\n", pr.get("permalink"), pr.get("name"), submitterName, formattedTimeUntilOverdue, prStatus));
            }
        }

        return messageText.toString();
    }

    public static void checkSla(Slack client) {
        LocalDateTime now = LocalDateTime.now();
        logger.info(String.format("Running SLA check at %s", now));
        int currentHour = now.getHour();
        List<Map<String, Object>> prStore = Database.getPrsFromStore();
        Map<String, Object> results = populateSla(prStore);

        List<String> prsToRemove = (List<String>) results.get("prsToRemove");
        Map<String, PriorityQueue<Map.Entry<Long, String>>> overduePrHeap = (Map<String, PriorityQueue<Map.Entry<Long, String>>>) results.get("overduePrHeap");
        Map<String, List<Map.Entry<Map<String, Object>, Long>>> nearSlaPrs = (Map<String, List<Map.Entry<Map<String, Object>, Long>>>) results.get("nearSlaPrs");

        for (String prId : prsToRemove) {
            Database.removePrById(prId);
        }

        Set<String> channelsToNotify = new HashSet<>(overduePrHeap.keySet());
        channelsToNotify.addAll(nearSlaPrs.keySet());

        for (String channelId : channelsToNotify) {
            List<Integer> enabledHours = DatabaseChannelSettings.getChannelEnabledHours(channelId);
            if (!enabledHours.contains(currentHour)) {
                logger.info(String.format("Skipping channel %s, current hour %d is not within enabled hours.", channelId, currentHour));
                continue;
            }

            String messageText = populateMessageText(channelId, overduePrHeap, nearSlaPrs);
            if (!messageText.isEmpty()) {
                try {
                    ChatPostMessageResponse response = client.methods().chatPostMessage(r -> r.channel(channelId).text(messageText).unfurlLinks(false));
                    logger.info(String.format("Sent SLA notification for channel %s.", channelId));
                } catch (Exception e) {
                    logger.error(String.format("Error sending SLA notification for channel %s: %s", channelId, e.getMessage()));
                }
            }
        }
    }

    public static void main(String[] args) {
        Slack client = Slack.getInstance();
        checkSla(client);
    }
}
